docs: [WIP] RFC for Optimistic Execution (#16499)
cosmos/cosmos-sdk: docs/rfc/rfc-002-optimistic-execution.md

> # RFC 002: Optimistic Execution

Reviewer: This number should change.

Author: Are we tracking what's the latest somewhere? Or should I check among open PRs?

Reviewer: It should be tracked here: https://docs.cosmos.network/main/rfc. Probably the table content of that page needs to be as well updated.

Reviewer 2: lets modify this to 5

Given that the application receives the block proposal in an earlier stage (`ProcessProposal`), it can be executed in the background so when `FinalizeBlock` is called the response can returned instantly.

## Decision

The newly introduced ABCI method `ProcessProposal` is called after a node receives the full block proposal of the current height but before prevote starts. CometBFT states that preemptively executing the block proposal is a potential use case for it:

> * **Usage**:
>     * Contains all information on the proposed block needed to fully execute it.
>         * The Application may fully execute the block as though it was handling
>          `RequestFinalizeBlock`.
>         * However, any resulting state changes must be kept as _candidate state_,
>           and the Application should be ready to discard it in case another block is decided.
>     * The Application MAY fully execute the block &mdash; immediate execution

Nevertheless, synchronously executing the proposal preemptively would not improve block time because it would just change the order of events (so the time we would like to save will be spent at `ProcessProposal` instead of `FinalizeBlock`).

> Instead we need to make block execution asynchronous by starting a goroutine in `ProcessProposal` (whose termination signal is kept in the application context) and returning a response immediately. That way, the actual block execution would happen at the same time as voting. When voting finishes and `FinalizeBlock` is called, the application handler can wait for the previously started goroutine to finish, and commit the resulting cache store if the block hash matches.

Reviewer 3: ```suggestion
Instead, we need to make block execution asynchronous by starting a goroutine in `ProcessProposal` (whose termination signal is kept in the application context) and returning a response immediately. That way, the actual block execution would happen at the same time as voting. When voting finishes and `FinalizeBlock` is called, the application handler can wait for the previously started goroutine to finish, and commit the resulting cache store if the block hash matches.
```

Assuming average voting period takes `P` and average block execution takes `Q`, this would reduce the average block time by `P + Q - max(P, Q)`.

Sei Network reported `P=~600ms` and `Q=~300ms` during a load test, meaning that optimistic execution could cut the block time by ~300ms.

The following diagram illustrates the intended flow:

```mermaid
flowchart LR;
    Propose-->ProcessProposal;
    ProcessProposal-->Prevote;
    Precommit-->FinalizeBlock;
    FinalizeBlock-->Commit;
    subgraph CometBFT
    Propose
    Prevote-->Precommit;
    Commit
    end
    subgraph CosmosSDK
    OEG[Optimistic Execution Goroutine]-.->FinalizeBlock
    ProcessProposal-.->OEG
    FinalizeBlock
    end
```

Some considerations:

- In the case that a proposal is being rejected by the local node, this node won't attempt to execute the proposal.
- The app must drop any previous Optimistic Execution data if `ProcessProposal` is called, in other words, abort and restart.

## Implementation

The execution context needs to have the following information:
- The block proposal (`abci.RequestFinalizeBlock`)
- Termination and completion signal for the OE goroutine

> The OE goroutine would run on top of a cached branch of the KVStore (which is the default behavior for `FinalizeBlock` as we only write to the underlying store once we've reached the end).

Reviewer 3: ```suggestion
The OE goroutine would run on top of a cached branch of the root multi-store (which is the default behavior for `FinalizeBlock` as we only write to the underlying store once we've reached the end).
```

The OE goroutine would periodically check if a termination signal has been sent to it, and stops if so. Once the OE goroutine finishes the execution it will set the completion signal.

> Upon receiving a `ProcessProposal` call, the SDK would adopt the following procedure if OE is enabled:

Reviewer 3: You mention "if OE is enabled" here, indicating that OE is optional. A few thoughts:

1. If we do indeed deem this as worthy of being optional, it should be mentioned earlier in the RFC IMO
2. What benefits do we gain from making this locally optional? If 50% of validators enable it and the remainder don't, then a network really won't see improvements in block time, will it? Is it worth making optional?

Author: I thought it would be useful in case of a malfunction, although I haven't really thought about it too much, I just defaulted to optional. But now that I think about it, yeah I can't come up with a real reason.
Maybe because we talked about the case of a block with a long running time having multiple rounds and I thought that it could be helpful to be able to disable OE. But making it optional wouldn't change a thing right?

Reviewer 3: I would actually argue against making this optional. I fear block times could sway too much depending on operator configurations.

Reviewer 2: maybe lets make this a parameter on NewBaseapp so that the application dev can decide if they want it or not. That way if something goes wrong there is a global way to revert this.

```

> abort any running OE goroutine

Reviewer 3: ```suggestion
abort any running OE goroutine and wait for goroutine exit
```

> if height > initial height:

Reviewer 3: Why is the `if height > initial height:` explicit check needed?

Reviewer 3: Nvm, I think you answered it below.

    set OE fields
    kick off an OP goroutine that optimistically process the proposal
else:
    do nothing
respond to CometBFT
```

OE won't be enabled during the first block of the chain, regardless of the configuration. This is because during the first block `InitChain` writes to `finalizeBlockState` so `FinalizeBlock` starts with an already initialized state. In the case that an abort would be needed it would mean we need to run `InitChain` again or discard/recover the state; either way complicates the implementation too much just for a single block.

Upon receiving a `FinalizeBlock` call, the SDK would adopt the following procedure:

```
if OE is enabled:
    abort if the currently executing OE's hash doesn't match the proposal's hash
    if aborted:
        process the proposal synchronously
    else:

>         respond to CometBFT

Reviewer 3: ...once the goroutine finishes
